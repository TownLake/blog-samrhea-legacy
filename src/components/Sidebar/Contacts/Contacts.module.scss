@import "src/assets/scss/variables";
@import "src/assets/scss/mixins";

.contacts {
  @include margin-bottom(1);

  .list {
    display: flex;
    flex-flow: row wrap;
    flex-grow: 0;
    flex-shrink: 0;
    list-style: none;
    max-width: 165px;
    @include padding-equal(0);
    @include margin(0.5, 0);

    .item {
      align-content: center;
      align-items: center;
      border: 1px solid var(--color-contacts-border);
      border-radius: 50%;
      display: flex;
      height: $button-height;
      justify-content: center;
      line-height: $button-height;
      text-align: center;
      width: $button-height;
      @include padding-equal(0);
      @include margin-equal(0.25);

      &:nth-child(3n + 1) {
        @include margin-left(0);
      }

      .link {
        border: 0;
        color: var(--color-typographic-base-font);
        cursor: pointer;
        display: flex;

        &:hover,
        &:focus {
          color: var(--color-typographic-link-p-font);
        }
      }
    }
  }
}
