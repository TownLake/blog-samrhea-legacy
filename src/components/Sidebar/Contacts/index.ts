export { default as Contacts } from "./Contacts";
