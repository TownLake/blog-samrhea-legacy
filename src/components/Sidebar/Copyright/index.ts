export { default as Copyright } from "./Copyright";
