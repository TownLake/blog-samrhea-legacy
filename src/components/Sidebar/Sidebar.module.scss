@import "../../assets/scss/mixins";

.sidebar {
  width: 100%;

  .inner {
    @include padding(1, 0.75, 0);
    position: relative;

    &:after {
      background: var(--color-sidebar-border);
      background: linear-gradient(
          to bottom,
          var(--color-sidebar-border) 0%,
          var(--color-sidebar-border) 48%,
          var(--color-sidebar-border-fade) 100%
      );
    }
  }
}

@include breakpoint-sm {
  .sidebar {
    lost-column: 5/12;

    .inner {
      @include padding(1.25, 0.75, 0);

      &:after {
        bottom: 0;
        content: "";
        height: 540px;
        position: absolute;
        right: -10px;
        top: 30px;
        width: 1px;
      }
    }
  }
}

@include breakpoint-md {
  .sidebar {
    lost-column: 1/3;

    .inner {
      @include padding-equal(1.5);
    }
  }
}
