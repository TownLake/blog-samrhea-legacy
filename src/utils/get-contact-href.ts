import { btoa } from "abab";

const getContactHref = (name: string, contact: string) => {
  const hrefs: { [key: string]: string } = {
    email: btoa(contact) || "",
    line: `line://ti/p/${contact}`,
    telegram: `https://t.me/${contact}`,
    vkontakte: `https://vk.com/${contact}`,
    medium: `https://medium.com/${contact}`,
    github: `https://github.com/${contact}`,
    weibo: `https://www.weibo.com/${contact}`,
    gitlab: `https://www.gitlab.com/${contact}`,
    codepen: `https://www.codepen.io/${contact}`,
    twitter: `https://www.twitter.com/${contact}`,
    soundcloud: `https://soundcloud.com/${contact}`,
    facebook: `https://www.facebook.com/${contact}`,
    instagram: `https://www.instagram.com/${contact}`,
    linkedin: `https://www.linkedin.com/in/${contact}`,
    youtube: `https://www.youtube.com/channel/${contact}`,
    mastodon: `${contact}`,
  };

  return hrefs[name] ?? contact;
};

export default getContactHref;
