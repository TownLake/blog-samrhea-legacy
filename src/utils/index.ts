export { default as getIcon } from "./get-icon";
export { default as getContactHref } from "./get-contact-href";
export { default as toKebabCase } from "./to-kebab-case";
export { default as getDefaultColorMode } from "./get-default-color-mode";
export * as testUtils from "./test-utils";
