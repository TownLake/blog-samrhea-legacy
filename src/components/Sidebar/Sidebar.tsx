import React from "react";

import { useSiteMetadata } from "@/hooks";

import { Author } from "./Author";
import { Contacts } from "./Contacts";
import { Copyright } from "./Copyright";
import { Menu } from "./Menu";

import * as styles from "./Sidebar.module.scss";

type Props = {
  isIndex?: boolean;
};

const Sidebar = ({ isIndex }: Props) => {
  const { author, copyright, menu } = useSiteMetadata();

  return (
    <div className={styles.sidebar}>
      <div className={styles.inner}>
        <Author author={author} isIndex={isIndex} />
        <Menu menu={menu} />
        <Contacts contacts={author.contacts} />
        <Copyright copyright={copyright} />
      </div>
    </div>
  );
};

export default Sidebar;
