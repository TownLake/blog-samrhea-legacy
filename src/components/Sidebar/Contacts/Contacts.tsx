import React from "react";

import { atob } from "abab";

import { Icon } from "@/components/Icon";
import { ICONS } from "@/constants";
import { Dictionary } from "@/types";
import { getContactHref, getIcon } from "@/utils";

import * as styles from "./Contacts.module.scss";

type Props = {
  contacts: Dictionary<string>;
};

const Contacts: React.FC<Props> = ({ contacts }: Props) => {
  const contactsEntered: Dictionary<string> = Object.keys(contacts).reduce(
    (acc, key) =>
      contacts[key] && contacts[key] !== "" && contacts[key] !== "#"
        ? { ...acc, [key]: contacts[key] }
        : acc,
    {},
  );

  return (
    Object.keys(contactsEntered).length > 0 && (
      <div className={styles.contacts}>
        <ul className={styles.list}>
          {(Object.keys(contactsEntered) as Array<keyof typeof ICONS>).map(
            (name) => (
              <li className={styles.item} key={name}>
                {name === "email" ? (
                  <span
                    className={styles.link}
                    onClick={() => {
                      window.location.href =
                        "mailto:" +
                        atob(getContactHref(name, contactsEntered[name]));
                    }}
                  >
                    <Icon name={name} icon={getIcon(name)} />
                  </span>
                ) : (
                  <a
                    className={styles.link}
                    href={getContactHref(name, contactsEntered[name])}
                    target="_blank"
                    rel={`noopener noreferrer${
                      name === "mastodon" ? " me" : ""
                    }`}
                  >
                    <Icon name={name} icon={getIcon(name)} />
                  </a>
                )}
              </li>
            ),
          )}
        </ul>
      </div>
    )
  );
};

export default Contacts;
