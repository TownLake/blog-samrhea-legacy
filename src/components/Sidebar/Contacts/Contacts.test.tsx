import React from "react";

import { Contacts } from "@/components/Sidebar/Contacts";
import * as mocks from "@/mocks";
import { testUtils } from "@/utils";

describe("Contacts", () => {
  test("renders null on empty contacts", () => {
    const props = { contacts: mocks.contacts };
    const tree = testUtils
      .createSnapshotsRenderer(<Contacts {...props} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test("renders a list of non-empty contacts", () => {
    const contacts = {
      ...mocks.contacts,
      github: "alxshelepenok/gatsby-starter-lumen",
    };
    const props = { contacts };
    const tree = testUtils
      .createSnapshotsRenderer(<Contacts {...props} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
