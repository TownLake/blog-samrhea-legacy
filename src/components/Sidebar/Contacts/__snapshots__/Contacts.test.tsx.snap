// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Contacts renders a list of non-empty contacts 1`] = `
<div>
  <ul>
    <li>
      <a
        href="https://github.com/alxshelepenok/gatsby-starter-lumen"
        rel="noopener noreferrer"
        target="_blank"
      >
        <svg
          viewBox="0 0 26 28"
        >
          <title>
            github
          </title>
          <path
            d="M10 19c0 1.141-0.594 3-2 3s-2-1.859-2-3 0.594-3 2-3 2 1.859 2 3zM20 19c0 1.141-0.594 3-2 3s-2-1.859-2-3 0.594-3 2-3 2 1.859 2 3zM22.5 19c0-2.391-1.453-4.5-4-4.5-1.031 0-2.016 0.187-3.047 0.328-0.812 0.125-1.625 0.172-2.453 0.172s-1.641-0.047-2.453-0.172c-1.016-0.141-2.016-0.328-3.047-0.328-2.547 0-4 2.109-4 4.5 0 4.781 4.375 5.516 8.188 5.516h2.625c3.813 0 8.188-0.734 8.188-5.516zM26 16.25c0 1.734-0.172 3.578-0.953 5.172-2.063 4.172-7.734 4.578-11.797 4.578-4.125 0-10.141-0.359-12.281-4.578-0.797-1.578-0.969-3.437-0.969-5.172 0-2.281 0.625-4.438 2.125-6.188-0.281-0.859-0.422-1.766-0.422-2.656 0-1.172 0.266-2.344 0.797-3.406 2.469 0 4.047 1.078 5.922 2.547 1.578-0.375 3.203-0.547 4.828-0.547 1.469 0 2.953 0.156 4.375 0.5 1.859-1.453 3.437-2.5 5.875-2.5 0.531 1.062 0.797 2.234 0.797 3.406 0 0.891-0.141 1.781-0.422 2.625 1.5 1.766 2.125 3.938 2.125 6.219z"
          />
        </svg>
      </a>
    </li>
  </ul>
</div>
`;

exports[`Contacts renders null on empty contacts 1`] = `null`;
